### DISCUSSION TITLE
keyboard navigation

### DISCUSSION THREAD

**Discussion by hollandjb92** [TanStack/table#2752] [General]
Wondering if there is any way to setup navigation of a table's rows through up/down arrow keys or is this something not available through react-table? getTrProps only works with onClick events

**Comment by GuptaSiddhant** [comment]
This feature is not natively supported but I was able to simulate it using following function:
Note: It is important that `tbody` tag should have a ref, and `tr` should have the `id={row.id}` and tabIndex

```jsx
...
  const tbodyRef = React.useRef(null);
  
 // Function to handle key press on a row
  const handleKeyDown = ( event, row ) => {
    event.stopPropagation();
    const currentRow = tbodyRef.current?.children.namedItem(row.id);
    switch (event.key) {
      case "ArrowUp": 
        currentRow?.previousElementSibling?.focus();
        break;      
      case "ArrowDown": 
        currentRow?.nextElementSibling?.focus();
        break;  
      default: break;
    }
  };
...
//Then find the function with TableRow `tr`
<tbody {...getTableBodyProps()} ref={tbodyRef}>
  {rows.map((row) => {
     prepareRow(row);
     return (
    <tr {...row.getRowProps()} id={row.id} tabIndex={0} onKeyDown={(e) => handleKeyDown(e, row)} >
    ...
   </tr>)}
</tbody>
...
```

To see selected rows, CSS must added to focus state of the row:
```css
tr:focus {
  background-colour: LightGray;
}
```

CodeSandBox: https://codesandbox.io/s/laughing-antonelli-yvsyl
Demo: https://yvsyl.csb.app

**Reply by danvln** [reply]
Thanks for sharing this sample, @GuptaSiddhant ! Any idea how to move between cells as well?

**Reply by GuptaSiddhant** [reply]
If you just add `tabIndex={0}` to each `td`, you'll be able to navigate with Tab. 
If you want arrowKey navigation, then you need to add similar event listeners to cells as I added to rows, but for left/right keys.

**Reply by danvln** [reply]
Thanks a lot for your quick reply. There is a problem because I am using virtualization and not all the rows and cells are on the view port. Any other idea that I can use?

**Reply by GuptaSiddhant** [reply]
I am sorry that I am not very well versed with virtualisation and its effects.
But if I had to do something in which I can't use tableRef to detect rows, I'd use react-table's rowSelect API or rowStateAPI to manually select/highlight the next or previous row. That would need to be calculated from current row Id. 
If you solve it, do share it. Good luck.

**Reply by dekokdj1** [reply]
On the subject of accessibility, is it possible to navigate the row headers that are used to sort with a keyboard?

**Reply by GuptaSiddhant** [reply]
Yes. It is. You need to add tabIndex to sort icons/buttons or on whole header cell, depending where you want user to click. You may need to write an custom onKeyPress method to perform sorting. 

There is a UI library built around react-table which does that. You can check it out. 
https://www.npmjs.com/package/react-table-ui

**Comment by bradwyatt** [comment]
Are there any future plans to support this natively? The closest solution I could get to this was by using the https://www.npmjs.com/package/accessible-react-table package, but it wasn't consistent across browsers, as well as the different pages on my app.

**Reply by GuptaSiddhant** [reply]
Unfortunately package `accessible-react-table` does not work with react-table 7 or TanStack table 8.

**Reply by Pagebakers** [reply]
I might be up for the task to implement this for v8. Will investigate.

**Reply by iamrakhmatov** [reply]
is there any news, regarding this?

**Comment by konsalex** [comment]
Hey folks, I recently released a library to work with table navigation. It works quite smoothly with `@tanstack/table`, and follows W3C guidelines for data grids. Feedback is more than welcome.

We use it in Neo4j Design system for table navigation, and works quite well for our use cases. 

If docs are confusing please ping me to help.

Library (npm): https://www.npmjs.com/package/@table-nav/react

Also example with [`@tanstack/table`](https://konsalex.github.io/table-nav/?path=/docs/tanstack-table-expanding-table--docs)

**Reply by Shank0071** [reply]
@konsalex This works great! But I am having a problem when I am dealing with inputs inside the table cell. "tab" key that is natively supported by the browsers does that work. However, I want to be able to navigate on "arrow key" events. Is there any way to achieve the same.

**Reply by konsalex** [reply]
Yes this issue [is known](https://github.com/konsalex/table-nav?tab=readme-ov-file#known-issues-), I consider it an issue with the spec to be honest, their is a thread also here but quite stale: https://github.com/w3c/aria-practices/issues/2739

I guess I could provide in the library a flag to disable cell navigation, but if you want to solve it for now it's easy.

On the events `onFocus` and `onBlur` you can disable/enable the keyboard navigation. [Check here the example also](https://github.com/konsalex/table-nav/blob/main/packages/storybook/stories/NeedleTable.tsx#L75)

**Reply by Shank0071** [reply]
Thank you for the quick response. I will take a look at the example!

**Comment by umkl** [comment]
In case anyone is looking for a really simple and quick solution for Tanstack Table, I created a custom hook that requires refs to each row:

```
import { keyboardMode } from "@/const/local-storage";
import { RowModel } from "@tanstack/react-table";
import { useEffect, useRef, useState } from "react";

export default function useKeyboardMode(
  rowModel: RowModel<YourTableType>,
) {
  const [keyboardEnabled, setKeyboardEnabled] = useState(
    localStorage.getItem(keyboardMode) === "enabled",
  );
  const rowRefs = useRef<(HTMLTableRowElement | null)[]>([]);

  const keyboardSelectionIdxRef = useRef(0);

  useEffect(() => {
    const handleKeyDown = (event: KeyboardEvent) => {
      if (!keyboardEnabled) {
        return;
      }
      const { key } = event;
      const totalRows = rowRefs.current.length;
      const prevIdx = keyboardSelectionIdxRef.current;
      let nextIdx = prevIdx;

      if (key === "ArrowDown") nextIdx = Math.min(prevIdx + 1, totalRows - 1);
      if (key === "ArrowUp") nextIdx = Math.max(prevIdx - 1, 0);
      if (key === " ") {
        const active = document.activeElement as HTMLElement | null;
        if (active?.tabIndex === 0) {
          return;
        } else {a
          rowModel.rows[keyboardSelectionIdxRef.current].toggleSelected();
        }
      }
      if (nextIdx !== prevIdx) {
        rowRefs.current[prevIdx]?.classList.remove("bg-green-800");
        rowRefs.current[nextIdx]?.classList.add("bg-green-800");
        rowRefs.current[nextIdx]?.scrollIntoView({
          behavior: "smooth",
          block: "center",
        });
        keyboardSelectionIdxRef.current = nextIdx;
      }
    };

    window.addEventListener("keydown", handleKeyDown);
    return () => window.removeEventListener("keydown", handleKeyDown);
  }, [keyboardEnabled, rowModel.rows]);

  return {
    keyboardEnabled,
    setKeyboardEnabled,
    rowRefs,
    keyboardSelectionIdxRef,
  };
}

```


And that is how you can consume the hook (keep in mind that this is a solution for Tanstack Table):

```
const { rowRefs, setKeyboardEnabled, keyboardSelectionIdxRef } =
    useKeyboardMode(table.getRowModel());
```


Make sure that there is a reference to each row in the using component:

```
<TableRow
                      key={row.id}
                      ref={(el) => {
                        rowRefs.current[idx] = el;
                      }}
```
